### ISSUE TITLE
2019-11-12 "Subscript out of bounds" for some partial AUCs

### ISSUE THREAD

**Issue by jhhughes256** [tensfeldt/openNCA#6]
### Error 
 
There are specific cases in commit a06c1c7461 where an error occurs when asking for partial AUCs. I am not certain why this error occurs for some partial AUCs and not others. The error and traceback for an example using tc010 has been included below. 
 
```
Error in value[[3L]](cond) : Error in tmp[[2]]: subscript out of bounds
For SDEID 63536292
In addition: Warning messages:
1: In run_M1_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTTAU' cannot be computed if 'TAUi' is not provided
2: In run_M1_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Error in value[[3L]](cond) : Error in tmp[[2]]: subscript out of bounds
For SDEID 63536292 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M1_SD_computation.R#1698
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    tmp_df <- data_data[data_data[, map_data$SDEID] == unique(data_data[, 
        map_data$SDEID])[i], ]
    tmp_df <- tmp_df[order(tmp_df[, map_data$TIME]), ] ... at run_M1_SD_computation.R#817
2. run_M1_SD_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2217
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset)
```

The error can be replicated by adding an additional partial AUC to test case 010 using the code below:

```
### Add partial AUC
mct$AUC.4.T1 <- 25
mct$AUC.4.T2 <- 48
```

The error does not occur when `mct$AUC.4.T1` is given a value of 23 or 24. `mct$AUC.4.T2` appears to have some impact on the error, changing the SDEID which causes the error. A value of 144 results in SDEID 63536292 causing the error, while a value of 48 results in SDEID 2925168 causing the error
 
### Reproducible Example
 
```
#### Final changes to test ----------------------------------------------- ####
### Add partial AUC
mct$AUC.4.T1 <- 25
mct$AUC.4.T2 <- 48

#### Run computation engine ---------------------------------------------- ####
# Run computation
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)
```

**Comment by jhhughes256** [comment]
### Update

As of commit e3be44d927 this error no longer occurs for the example listed above, however, it still occurs in other tests I have run. 

The error is the same, but appears to trace back to a different `tryCatch`, so may be the same error in a different place, or a completely different error that reads out the same way.

### Error and Traceback

#### tc014

Base test case with a partial AUC from 48 to 96 added. Error also occurs (with a different SDEID) when using a partial AUC from 48 to 240.

```
 Error in value[[3L]](cond) : Error in tmp[[2]]: subscript out of bounds
For SDEID 50896291 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M2_SS_computation.R#2296
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    if (comp_required[["C0"]]) {
        c_0 <- list()
    } ... at run_M2_SS_computation.R#935
2. run_M2_SS_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2235
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset)
```

#### tc016

Base test case with a partial AUC from 48 to 192 added. Error also occurs (with a different SDEID) when using a partial AUC from 48 to 63.

```
 Error in value[[3L]](cond) : Error in tmp[[2]]: subscript out of bounds
For SDEID 910511415 
7. stop(paste0(e, "For SDEID ", unique(data_data[, map_data$SDEID])[i])) at run_M3_SS_computation.R#2179
6. value[[3L]](cond) 
5. tryCatchOne(expr, names, parentenv, handlers[[1L]]) 
4. tryCatchList(expr, classes, parentenv, handlers) 
3. tryCatch({
    if (comp_required[["CMAXi"]]) {
        c_maxi <- list()
    } ... at run_M3_SS_computation.R#922
2. run_M3_SS_computation(data = merged_data, map = map_data, method = method, 
    model_regex = model_regex, parameter_list = parameter_list, 
    return_list = return_list) at run_computation.R#2247
1. run_computation(data = d, map = mct, flag = flags, parameterset = parameterset)
```

### Reproducible Example

#### tc014

```
#### Load Libraries ------------------------------------------------------ ####
library(readr)
library(jsonlite)
library(testthat)

#### Load specific code-base for OpenNCA --------------------------------- ####
# pkgload::load_all(path = "codebase/openNCA-master/openNCA")
# pkgload::load_all(path = "codebase/openNCA-b1977923fc/openNCA")  # 11/8-v1
# pkgload::load_all(path = "codebase/openNCA-051805c887/openNCA")  # 11/8-v2
# pkgload::load_all(path = "codebase/openNCA-1fc479c2ca/openNCA")  # 11/11
# pkgload::load_all(path = "codebase/openNCA-a06c1c7461/openNCA")  # 11/12
pkgload::load_all(path = "codebase/openNCA-e3be44d927/openNCA")  # 11/13

#### Clean test environment ---------------------------------------------- ####
# Clean environment
rm(params, conc, vlist, results, results_list, r, e, mct, d, flags, p, klist, n, 
  ninc, nexc, k, rlist, dlist, nlist, slist, df, xmin, xmax, dk, flag_data, 
  est_data, outputdir)
 
profileid <- "SDEID"
args = commandArgs(trailingOnly=TRUE)
params = c()
 
#### Test Case 014 ------------------------------------------------------- ####
# M2 SS - nominal time
# p <- "C:/Users/tensfeldt/Box Sync/EQuIP Computation Engine/Rudraya_executed_test_cases/tc014"
testcase.id <- "tc014"
p <- "testcase/tc014_M2SS"
params["INPUT1"] = file.path(p, "A4021010.in.csv")
params["parameterset"] ="PARAMETERLIST"
params["map"] = file.path(p, "tc014_mct_updated.csv")
params["flags"] = file.path(p, "A4021010_flag_data.in.csv")
params["OUTPUT"] ="."
print(params)
 
mct <- read_csv(params["map"])
write_file(toJSON(mct, na="string"), path=file.path(p, "tc014_mct_updated.json"))
params["map"] = file.path(p, "tc014_mct_updated.json")
 
flags <- read_csv(params["flags"])
write_file(toJSON(flags, na="string"), path=file.path(p, "tc014_flags.json"))
params["flags"] = file.path(p, "tc014_flags.json")
 
#### Read full concentration dataset ------------------------------------- ####
# Read conc
conc <- read.csv(file=params["INPUT1"], header=TRUE, sep=",", as.is=TRUE)
vlist <- c("PKDATAROWID", "PKPTMS", "PKCNCN", "PERIOD", "TREATXT")
 
# Read mct
mct <- params["map"]
###mct <- read_file(mct)
###mct <- gsub("[\\]+", "", mct)
mct <- fromJSON(mct)
 
# Read flags
flags <- params["flags"]
###flags <- read_file(flags)
###flags <- gsub("[\\]+", "", flags)
flags <- fromJSON(flags)
 
# ParameterSet
if(is.na(params["parameterset"])) { params["parameterset"] <- "PARAMETERDISPLAYLIST" }
parameterset <- params["parameterset"]
 
# FLAGS is a subset for the specific profile, subset concentration dataframe 
#  to those records by PKDATAROWID
d <- conc[is.element(conc$PKDATAROWID, flags$PKDATAROWID),]
 
# Add in missing required parameters for virtual compute call
klist <- c("KEL", "KELRSQ", "KELNOPT", "THALFF")
if(!is.element("PARAMETERDISPLAYLIST", names(mct))) { mct$PARAMETERDISPLAYLIST <- "" }
for(i in klist) {
  ###if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- c(mct$PARAMETERDISPLAYLIST, i) }
  if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- paste(mct$PARAMETERDISPLAYLIST, i, sep=";") }
}
 
# Remove any leading semi-colons ";"
mct$PARAMETERDISPLAYLIST <- gsub("^;*", "", mct$PARAMETERDISPLAYLIST, ignore.case=TRUE, perl=TRUE)
 
# For virtual compute call only: Subset the parameters to be calculated to those in parameterdisplaylist
if(!is.element("PARAMETERLIST", names(mct))) { mct$PARAMETERLIST <- "" }
 
# Check whether TIME selection has been set in MCT - default to NOMINAL
if(!is.element("TIME", names(mct))) { mct$TIME <- "NOMINAL" }
 
# Determine whether there are any points selected for inclusion in the terminal elimination phase via 
#  FLGEXKEL (0=not excluded/included; 1=excluded/not included)
#  ***What happens if there are no flags provided?***
n <- factor(flags$FLGEXKEL, levels=c(0,1))
ninc <- length(n[n==0])
nexc <- length(n[n==1])
 
#### Final changes to test ----------------------------------------------- ####
### Add partial AUC
mct$AUC.1.T1 <- 48
mct$AUC.1.T2 <- 96
 
#### Run computation engine ---------------------------------------------- ####
# Run computation
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)
```

#### tc016

```
#### Load Libraries ------------------------------------------------------ ####
library(readr)
library(jsonlite)
library(testthat)

#### Load specific code-base for OpenNCA --------------------------------- ####
# pkgload::load_all(path = "codebase/openNCA-master/openNCA")
# pkgload::load_all(path = "codebase/openNCA-b1977923fc/openNCA")  # 11/8-v1
# pkgload::load_all(path = "codebase/openNCA-051805c887/openNCA")  # 11/8-v2
# pkgload::load_all(path = "codebase/openNCA-1fc479c2ca/openNCA")  # 11/11
# pkgload::load_all(path = "codebase/openNCA-a06c1c7461/openNCA")  # 11/12
pkgload::load_all(path = "codebase/openNCA-e3be44d927/openNCA")  # 11/13

#### Clean test environment ---------------------------------------------- ####
# Clean environment
rm(params, conc, vlist, results, results_list, r, e, mct, d, flags, p, klist, n, 
  ninc, nexc, k, rlist, dlist, nlist, slist, df, xmin, xmax, dk, flag_data, 
  est_data, outputdir)
 
profileid <- "SDEID"
args = commandArgs(trailingOnly=TRUE)
params = c()
 
#### Test Case 016 ------------------------------------------------------- ####
# M3 SS
# p <- "C:/Users/tensfeldt/Box Sync/EQuIP Computation Engine/Rudraya_executed_test_cases/tc016"
testcase.id <- "tc016"
p <- "testcase/tc016_M3SS"
params["INPUT1"] = file.path(p, "B2901002.in.csv")
params["parameterset"] ="PARAMETERLIST"
params["map"] = file.path(p, "tc016_mct_updated.csv")
params["flags"] = file.path(p, "B2901002_flag_data.in.csv")
params["OUTPUT"] ="."
print(params)
 
mct <- read_csv(params["map"])
write_file(toJSON(mct, na="string"), path=file.path(p, "tc016_mct_updated.json"))
params["map"] = file.path(p, "tc016_mct_updated.json")
 
flags <- read_csv(params["flags"])
write_file(toJSON(flags, na="string"), path=file.path(p, "tc016_flags.json"))
params["flags"] = file.path(p, "tc016_flags.json")
 
#### Read full concentration dataset ------------------------------------- ####
# Read conc
conc <- read.csv(file=params["INPUT1"], header=TRUE, sep=",", as.is=TRUE)
vlist <- c("PKDATAROWID", "PKPTMS", "PKCNCN", "PERIOD", "TREATXT")
 
# Read mct
mct <- params["map"]
###mct <- read_file(mct)
###mct <- gsub("[\\]+", "", mct)
mct <- fromJSON(mct)
 
# Read flags
flags <- params["flags"]
###flags <- read_file(flags)
###flags <- gsub("[\\]+", "", flags)
flags <- fromJSON(flags)
 
# ParameterSet
if(is.na(params["parameterset"])) { params["parameterset"] <- "PARAMETERDISPLAYLIST" }
parameterset <- params["parameterset"]
 
# FLAGS is a subset for the specific profile, subset concentration dataframe 
#  to those records by PKDATAROWID
d <- conc[is.element(conc$PKDATAROWID, flags$PKDATAROWID),]
 
# Add in missing required parameters for virtual compute call
klist <- c("KEL", "KELRSQ", "KELNOPT", "THALFF")
if(!is.element("PARAMETERDISPLAYLIST", names(mct))) { mct$PARAMETERDISPLAYLIST <- "" }
for(i in klist) {
  ###if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- c(mct$PARAMETERDISPLAYLIST, i) }
  if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- paste(mct$PARAMETERDISPLAYLIST, i, sep=";") }
}
 
# Remove any leading semi-colons ";"
mct$PARAMETERDISPLAYLIST <- gsub("^;*", "", mct$PARAMETERDISPLAYLIST, ignore.case=TRUE, perl=TRUE)
 
# For virtual compute call only: Subset the parameters to be calculated to those in parameterdisplaylist
if(!is.element("PARAMETERLIST", names(mct))) { mct$PARAMETERLIST <- "" }
 
# Check whether TIME selection has been set in MCT - default to NOMINAL
if(!is.element("TIME", names(mct))) { mct$TIME <- "NOMINAL" }
 
# Determine whether there are any points selected for inclusion in the terminal elimination phase via 
#  FLGEXKEL (0=not excluded/included; 1=excluded/not included)
#  ***What happens if there are no flags provided?***
n <- factor(flags$FLGEXKEL, levels=c(0,1))
ninc <- length(n[n==0])
nexc <- length(n[n==1])
 
#### Final changes to test ----------------------------------------------- ####
### Add partial AUC
mct$AUC.1.T1 <- 48
mct$AUC.1.T2 <- 192
 
#### Run computation engine ---------------------------------------------- ####
# Run computation
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)
```

**Comment by jhhughes256** [comment]
### Update

Additional testing shows that this issue occurs mainly when using an interpolated value in the partial AUC. An exception is when asking for a partial AUC from zero to an interpolated value. 

This error also occurs when asking for a partial AUC from an extrapolated value to another extrapolated value.

**Comment by jhhughes256** [comment]
### Update

This issue appears to be fixed for partial AUCs including:

- zero to existing value
- zero to interpolated value
- zero to extrapolated value
- existing value to existing value
- existing value to interpolated value
- existing value to extrapolated value
- interpolated value to existing value
- interpolated value to interpolated value
- interpolated value to extrapolated value

The only use case that returns this error is extrapolated value to extrapolated value, which may not be a likely request in practice. Therefore, I am removing the priority label from this issue and am happy for the issue to close pending input from @tensfeldt .

**Comment by jhhughes256** [comment]
### Update

Extrapolation to extrapolation partial AUCs have been discussed and will likely need to be addressed due to related issues.

**Comment by jhhughes256** [comment]
### Update

This has been fixed as of 02b2470a64. Partial AUCs for extrapolated to extrapolated points no longer cause a computation error.
